### PR TITLE
Introduce weak references

### LINKED ISSUE #48: SqliteDict memory issue
Building LSI model from SqliteDict corpus loads the whole SqliteDict into memory.

How to simulate:

``` python
from sqlitedict import SqliteDict
from gensim.models import TfidfModel, LsiModel
from gensim.models.lsimodel import Projection
from gensim.utils import grouper
from gensim import matutils

s = SqliteDict('documents.sqlite')  # contains namedtuples with field `tokens` containing unicode token tuple
tfidf = TfidfModel.load('tfidf_model.pkl')  # trained on a dictionary built from `s`


class SegmentCorpus(object):
    def __init__(self, storage, built_dictionary):
        self.storage = storage
        self.inner_dictionary = built_dictionary

    def __iter__(self):
        for segment in self.storage.itervalues():
            yield self.inner_dictionary.doc2bow(segment.tokens)


corpus_tfidf = tfidf[SegmentCorpus(s, tfidf.id2word)]
for i in grouper(corpus_tfidf, 5000):
    nnz = sum(len(doc) for doc in i)
    job = matutils.corpus2csc(i, num_docs=len(i), num_nnz=nnz, num_terms=len(tfidf.id2word))
    Projection(len(tfidf.id2word), 100, job, extra_dims=100, power_iters=1)
```

Calling Projection will increase RAM usage to the size of SqliteDict db (in my case 22GB).

### LINKED ISSUE #69: Interrupted iteration over a SqliteDict instance
I have a file created by sqlitedict in python2.7, file size is 11.37GB (number of keys:761951).
Recently I opened it with read-only flag in Python3.6. Something interesting:
Scenario A: If it is iterated over all  761951 keys (for key,value in dict.items()), the memory consumption is very low. Like this:
image

Scenario B: if the iteration was break prematurely, say at iteration counter=100, then call dict.close(). the memory consumption raises high. Like this:
image

Scenario C: if the iteration was break prematurely, say at iteration counter=100, but not to call dict.close(). The memory consumption keeps growing over 10 seconds after the execution has came out from the iteration. the memory consumption raises high as Scenario B. 

(All the Scenario A,B,C are running in a thread by twisted "reactor.callInThread")

It suggests that the iteration was still undergoing even it breaks.

### PR DESCRIPTION
**Pull request by mpenkov**
- Fix #69 
- Fix #48 (maybe)

Author is @Cologler

Continued from #153 because that branch was mistakenly deleted.

### CODE REVIEW

**Comment by mpenkov** [comment]
@piskvorky @Cologler Please eyeball this PR and let me know if you spot anything odd

**Comment by Cologler** [comment]
> @piskvorky @Cologler Please eyeball this PR and let me know if you spot anything odd

Looks Good To Me 👌.

File: sqlitedict.py [hunk 1]
```diff
@@ -67,6 +68,51 @@ def reraise(tp, value, tb=None):
...
+# There's a thread that holds the actual SQL connection (SqliteMultithread).
+# We communicate with this thread via queues (request and responses).
+# The requests can either be SQL commands or one of the "special" commands
+# below:
+#
+# _REQUEST_CLOSE: request that the SQL connection be closed
+# _REQUEST_COMMIT: request that any changes be committed to the DB
+#
+# Responses are either SQL records (e.g. results of a SELECT) or the magic
+# _RESPONSES_NO_MORE command, which indicates nothing else will ever be written
```
Comment (piskvorky) [inline sqlitedict.py:81]:
```suggestion
# _RESPONSE_NO_MORE command, which indicates nothing else will ever be written
```

**Review by piskvorky — CHANGES_REQUESTED** [review]

**Comment by Cologler** [comment]
What if the user does some longtime work in the loop?

``` py
for value in a_sqlite_dict.values():
    # long timework here, like an HTTP request
    ...
```

**Comment by mpenkov** [comment]
> What if the user does some longtime work in the loop?
> 
> ```python
> for value in a_sqlite_dict.values():
>     # longtime work here, like an HTTP request
>     ...
> ```

I think the response queue will continue to receive new items.

Basically, there is a [producer/consumer](https://en.wikipedia.org/wiki/Producer%E2%80%93consumer_problem) problem here. The producer is the sqlite thread and the consumer is the user. We want to produce results quickly enough for the user to consume them without overwhelming the user. If we produce too quickly (or the user is too slow to consume) then the queue grows in size and occupies more memory.

I think one way to solve that particular problem is to use fixed-size queues for the responses. If a queue is full, then it means the user is consuming faster than we are producing, and we can produce slower.

@piskvorky What do you think? Is there a reason why we use queues of unlimited size?

**Comment by piskvorky** [comment]
> If a queue is full, then it means the user is consuming faster than we are producing, and we can produce slower.

Do you mean "consuming slower"?

Anyway, limiting the queue is technically easy, here:

https://github.com/RaRe-Technologies/sqlitedict/blob/8545a174a4a32bd2f8eb7c17d89899370138fe3a/sqlitedict.py#L623-L624

But I'm not sure about the implications. The code is sufficiently complex that blocking `put`s could lead to dead locks, unless we think through all the code paths very carefully. What happens when the consumers don't `get` (for whatever reason – generator, exception…), so that the `put` blocks forever?

**Comment by mpenkov** [comment]
> Do you mean "consuming slower"?

Yes, sorry.

> Anyway, limiting the queue is technically easy, here:

Yes, as you point out, this is the easy part. The hard part is ensuring nothing else breaks.

> What happens when the consumers don't get (for whatever reason – generator, exception…), so that the put blocks forever?

Good question...

Anyway, I think this discussion is out of scope for this PR (but definitely helpful). I think we're done for now.

**Pull request merged**